File: java/src/main/java/io/fivexer/sdk/Skills.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.PatchSkill;
import io.fivexer.sdk.model.Skill;
import io.fivexer.sdk.model.SkillList;
import io.fivexer.sdk.model.UpsertSkill;
import java.util.Collections;
import java.util.List;

/** The workspace skill catalogue. */
public final class Skills {

    private final Fivexer client;

    Skills(Fivexer client) {
        this.client = client;
    }

    public Skill create(UpsertSkill input) {
        return client.request("POST", "/skills", input.toJson(), null, Skill.class, false);
    }

    public List<Skill> list() {
        return list(null, null);
    }

    /** @param q free-text filter over key and name */
    public List<Skill> list(String q, Integer limit) {
        SkillList envelope = client.request("GET", "/skills", null,
                Json.query("q", q, "limit", Json.str(limit)), SkillList.class, false);
        return unwrap(envelope);
    }

    public Skill get(String skillId) {
        return client.request("GET", "/skills/" + Json.enc(skillId), null, null, Skill.class, false);
    }

    /** Partial update — the {@code key} is immutable. */
    public Skill patch(String skillId, PatchSkill patch) {
        return client.request("PATCH", "/skills/" + Json.enc(skillId), patch.toJson(), null,
                Skill.class, false);
    }

    public void remove(String skillId) {
        client.request("DELETE", "/skills/" + Json.enc(skillId), null, null, Void.class, true);
    }

    /**
     * Skills commonly held alongside the ones already chosen. The list travels as one
     * comma-joined query parameter; an empty selection omits it entirely.
     */
    public List<Skill> suggest(List<String> selected, Integer limit) {
        String joined = (selected == null || selected.isEmpty()) ? null : String.join(",", selected);
        SkillList envelope = client.request("GET", "/skills/suggest", null,
                Json.query("selected", joined, "limit", Json.str(limit)), SkillList.class, false);
        return unwrap(envelope);
    }

    public List<Skill> suggest() {
        return suggest(null, null);
    }

    private static List<Skill> unwrap(SkillList envelope) {
        return envelope.getSkills() == null ? Collections.emptyList() : envelope.getSkills();
    }
}

File: php/src/Resource/Skills.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\PatchSkill;
use Fivexer\SDK\Model\Skill;
use Fivexer\SDK\Model\UpsertSkill;

/** The workspace skill catalogue. */
final class Skills
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    public function create(UpsertSkill $input): Skill
    {
        return Skill::fromArray($this->client->request('POST', '/skills', $input->toArray()) ?? []);
    }

    /**
     * @param string|null $q Free-text filter over key and name
     * @return list<Skill>
     */
    public function list(?string $q = null, ?int $limit = null): array
    {
        $data = $this->client->request('GET', '/skills', null, ['q' => $q, 'limit' => $limit]) ?? [];
        /** @var list<Skill> */
        return Json::parseEach($data, 'skills', [Skill::class, 'fromArray']);
    }

    public function get(string $skillId): Skill
    {
        return Skill::fromArray($this->client->request('GET', '/skills/' . \rawurlencode($skillId)) ?? []);
    }

    /** Partial update — the key is immutable. */
    public function patch(string $skillId, PatchSkill $patch): Skill
    {
        return Skill::fromArray(
            $this->client->request('PATCH', '/skills/' . \rawurlencode($skillId), $patch->toArray()) ?? []
        );
    }

    public function remove(string $skillId): void
    {
        $this->client->request('DELETE', '/skills/' . \rawurlencode($skillId));
    }

    /**
     * Skills commonly held alongside the ones already chosen. The list travels as one
     * comma-joined query parameter; an empty selection omits it entirely.
     *
     * @param list<string>|null $selected
     * @return list<Skill>
     */
    public function suggest(?array $selected = null, ?int $limit = null): array
    {
        $joined = ($selected === null || $selected === []) ? null : \implode(',', $selected);
        $data = $this->client->request('GET', '/skills/suggest', null, [
            'selected' => $joined,
            'limit' => $limit,
        ]) ?? [];
        /** @var list<Skill> */
        return Json::parseEach($data, 'skills', [Skill::class, 'fromArray']);
    }
}
